stellargraph/layer/knowledge_graph.py

```diff
@@ -191,3 +191,126 @@ def build(self):
         x_out = self(x_inp)
 
         return x_inp, x_out
+
+
+class DistMultScore(Layer):
+    """
+    DistMult scoring Keras layer.
+
+    Original Paper: Embedding Entities and Relations for Learning and Inference in Knowledge
+    Bases. Bishan Yang, Wen-tau Yih, Xiaodong He, Jianfeng Gao, Li Deng. ICLR 2015
+
+    This combines subject, relation and object embeddings into a score of the likelihood of the
+    link.
+    """
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+    def build(self, input_shape):
+        self.built = True
+
+    def call(self, inputs):
+        e1, r, e2 = inputs
+        # y_(e_1)^T M_r y_(e_2), where M_r = diag(w_r) is a diagonal matrix
+        score = tf.reduce_sum(e1 * r * e2, axis=2)
+        return score
+
+
+@experimental(reason="results from the reference paper have not been reproduced yet")
+class DistMult:
+    """
+    Embedding layers and a DistMult scoring layers that implement the DistMult knowledge graph
+    embedding algorithm as in https://arxiv.org/pdf/1412.6575.pdf
+
+    Args:
+        generator (KGTripleGenerator): A generator of triples to feed into the model.
+
+        k (int): the dimension of the embedding (that is, a vector in R^k is learnt for each node
+            and each link type)
+
+        embedding_initializer (str or func, optional): The initialiser to use for the embeddings.
+
+        embedding_regularizer (str or func, optional): The regularizer to use for the embeddings.
+    """
+
+    def __init__(
+        self, generator, k, embedding_initializer=None, embedding_regularizer=None,
+    ):
+        if not isinstance(generator, KGTripleGenerator):
+            raise TypeError(
+                f"generator: expected KGTripleGenerator, found {type(generator).__name__}"
+            )
+
+        graph = generator.G
+        self.num_nodes = graph.number_of_nodes()
+        self.num_edge_types = len(graph._edges.types)
+        self.k = k
```

Comment:
there's a lot of `k`s in the library; could you change this to something more descriptive?

```suggestion
        self.embedding_dimension = embedding_dimension
```

(with the necessary change to the input params)

Reply:
Much better.